Debug help for failed Plutus scripts

JaredCorduan opened the pull request:
This adds all the necessary information for re-runing a failed plutus script inside the `ValidationTagMismatch` predicate failure.

Additionally, I've replaced the `[Text]` field in `ValidationTagMismatch` with descriptive types for all the different scenarios.  Previously, the `Eq` instance for `ValidationTagMismatch` was ignoring this field, I suspect to make testing easier. I've made the `Eq` instance care about all the fields, and gave the tests a way to ignore the inner details of the plutus error message.

Commit:
Debug help for failed plutus scripts

This adds the hex encoding of the plutus inputs to the predicate failure
associated with failed scripts. It also adds a function 'debugPlutus'
which can be used to re-run a Plutus script in verbose mode.

michaelpj reviewed (approved):
Seems fine. We could move the executable into `plutus` if we wanted: we have an executable that does this sort of thing, it's only not set up to take its arguments as a single serialized blob. But here is fine too.

Commit:
separate plutus debug info in predicate failure

JaredCorduan commented:
@michaelpj I've removed your review since I've taken it out of draft and changed it a lot. Don't feel obligated to review it again (unless you want to), I can get a review from the usual suspects.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs:
```diff
@@ -237,11 +242,27 @@ scriptsNotValidateTransition = do
+data TagMismatchDescription
+  = PassedUnexpectedly
+  | FailedUnexpectedly [FailureDescription]
+  deriving (Show, Eq, Generic, NoThunks)
```
Deriving `NoThunks` here is a bit weird, since it's not obvious that we _will_ have no thunks. Are we forcing the descriptions during construction?

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -362,7 +381,35 @@ valContext ::
-data ScriptResult = Passes | Fails ![String]
+data FailureDescription
+  = OnePhaseFailure Text
+  | PlutusFailure Text ByteString
```
Label fields here?

JaredCorduan replied:
do you mean add some comments? that seems a good idea to me, thanks

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -362,7 +381,35 @@ valContext ::
+      dec 1 = SumD PlutusFailure <! From <! From
+      dec n = Invalid n
+
+data ScriptResult = Passes | Fails ![FailureDescription]
```
This bang will only force the head of the list...

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -395,81 +488,102 @@ runPLCScript proxy (CostModel cost) scriptbytestring units ds =
+                    "The third data argument, does not decode to a context\n" ++ show info
+                  ]
+            db = B64.encode . serialize' $ PlutusDebug cm eu scriptbytestring ds PlutusV1
+        Just info2 -> Fails [PlutusFailure line db]
```
`line` and `db` will be thunks here...

nc6 reviewed (approved):
So there's some weird stuff potentially with evaluation in this PR, and `NoThunks` checks on types that will be decidedly thunky. But none of that should change semantics, and in the name of getting a RC today, I'll merge and we can update this later.

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -395,81 +488,102 @@ runPLCScript proxy (CostModel cost) scriptbytestring units ds =
+                  [ "\nThe 2 arg plutus script (" ++ name ++ ") fails.",
+                    show e,
+                    "The redeemer is: " ++ show redeemer,
+                    "The second data argument, does not decode to a context\n" ++ show info
```
Is the textual information here now redundant with the debug information?

JaredCorduan replied:
yep, it is. maybe the human readable part should be whittled down, good point, thanks.

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -395,81 +488,102 @@ runPLCScript proxy (CostModel cost) scriptbytestring units ds =
       name = show ss
    in case P.fromData info of
-        Nothing -> Fails [line]
+        Nothing -> Fails [PlutusFailure line db]
```
also I forget, but could we put the structured type in here and just print it specially later?

JaredCorduan replied:
I think the train has left the station, sorry. But as for you question, I don't know, we would have to write our own show for it to produce the base64 in the logs messages, which is maybe a weird/not-always-what-you-want show instance?

michaelpj replied:
You're using `Show` for log messages? I'm not totally sure that's a good idea, but if that's the situation then okay.

JaredCorduan replied:
I'm not actually sure, to be honest.
